code/controllers/subsystem/demo.dm

```diff
@@ -0,0 +1,425 @@
+			var/client/C = CLIENT_FROM_VAR(T)
+			if(C)
+				target_keys += C.ckey
+		if(!target_keys.len)
+			return
+		target_text = jointext(target_keys, ",")
+	else
+		var/client/C = CLIENT_FROM_VAR(target)
+		if(C)
+			target_text = C.ckey
+		else
+			return
+	write_event_line("chat [target_text] [last_chat_message == text ? "=" : json_encode(text)]")
+	last_chat_message = text
+
+/datum/controller/subsystem/demo/Initialize()
+	WRITE_LOG_NO_FORMAT(GLOB.demo_log, "demo version 1\n") // increment this if you change the format
+	if(GLOB.revdata)
+		WRITE_LOG_NO_FORMAT(GLOB.demo_log, "commit [GLOB.revdata.originmastercommit || GLOB.revdata.commit]\n")
+
+	// write a "snapshot" of the world at this point.
+	// start with turfs
+	log_world("Writing turfs...")
+	WRITE_LOG_NO_FORMAT(GLOB.demo_log, "init [world.maxx] [world.maxy] [world.maxz]\n")
+	marked_turfs.len = 0
+	for(var/z in 1 to world.maxz)
+		var/row_list = list()
+		var/last_appearance
+		var/rle_count = 1
+		for(var/y in 1 to world.maxy)
+			for(var/x in 1 to world.maxx)
+				var/turf/T = locate(x,y,z)
+				T.demo_last_appearance = T.appearance
+				var/this_appearance
+				// space turfs are difficult to RLE otherwise, because they all
+				// have different appearances despite being the same thing.
+				if(T.type == /turf/open/space || T.type == /turf/open/space/basic)
+					this_appearance = "s" // save the bytes
+				else if(istype(T, /turf/open/space/transit))
+					this_appearance = "t[T.dir]"
+				else
+					this_appearance = T.appearance
+				if(this_appearance == last_appearance)
+					rle_count++
+				else
+					if(rle_count > 1)
+						row_list += rle_count
+					rle_count = 1
+					if(istext(this_appearance))
+						row_list += this_appearance
+					else
+						// do a diff with the previous turf to save those bytes
+						row_list += encode_appearance(this_appearance, istext(last_appearance) ? null : last_appearance)
+				last_appearance = this_appearance
+		if(rle_count > 1)
+			row_list += rle_count
+		WRITE_LOG_NO_FORMAT(GLOB.demo_log, jointext(row_list, ",") + "\n")
+	CHECK_TICK
+	// then do objects
+	log_world("Writing objects")
+	marked_new.len = 0
+	marked_dirty.len = 0
+	for(var/z in 1 to world.maxz)
+		var/spacing = 0
+		var/row_list = list()
+		for(var/y in 1 to world.maxy)
+			for(var/x in 1 to world.maxx)
+				var/turf/T = locate(x,y,z)
+				var/list/turf_list = list()
+				for(var/C in T.contents)
+					var/atom/movable/as_movable = C
+					if(as_movable.loc != T)
+						continue
+					if(isobj(C) || ismob(C))
+						turf_list += encode_init_obj(C)
+				if(turf_list.len)
+					if(spacing)
+						row_list += spacing
+						spacing = 0
+					row_list += turf_list
+				spacing++
+			CHECK_TICK // This is a bit risky because something might change but meh, its not a big deal.
+		WRITE_LOG_NO_FORMAT(GLOB.demo_log, jointext(row_list, ",") + "\n")
+
+	// track objects that exist in nullspace
+	var/nullspace_list = list()
+	for(var/atom/movable/M in world)
```

Comment:
not sure if this even needs to be a typed loop.

Reply:
pretty sure `var/datum/d in world` does something

Reply:
![worksfine](https://user-images.githubusercontent.com/47324920/71784100-d4acb900-2fef-11ea-8262-e563c8fa22a0.PNG)

Reply:
does it include turfs?

Reply:
yes

Reply:
Seems like it made it initialize like a half second faster while still producing working demos.